docs/tutorials/data_sources/huggingface_dataset_tutorial.md

```diff
@@ -0,0 +1,144 @@
+import numpy as np
+
+# HF imports
+from datasets import load_dataset
+from transformers import AutoTokenizer
+```
+
+Let's first import an HF dataset. For the sake of simplicity let's proceed with [lhoestq/demo1](https://huggingface.co/datasets/lhoestq/demo1) - a minimal dataset comprised of five rows and six columns.
+
+```{code-cell} ipython3
+hf_dataset = load_dataset("lhoestq/demo1")
+hf_train, hf_test = hf_dataset["train"], hf_dataset["test"]
+hf_dataset
+```
+
+Each sample is a Python dictionary with string or integer data.
+
+```{code-cell} ipython3
+hf_train[0]
+```
+
+Let's assume that for our preprocessing pipeline we want the string `date` field to become a timestamp and the whole sample - a NumPy array.
+
+```{code-cell} ipython3
+def process_date(sample: dict) -> dict:
+    sample["date"] = parse(sample["date"]).timestamp()
+    return sample
+
+def process_sample_to_np(sample: dict) -> np.ndarray:
+    return np.array([*sample.values()], dtype=object)
+```
+
+Building a pipeline is as simple as chaining `map` calls. HF dataset supports random access so we can pass it directly to a `source` method. The resulting object is of type `grain.MapDataset` with random access support.
+
+```{code-cell} ipython3
+dataset = (
+    grain.MapDataset.source(hf_train)
+    .shuffle(seed=10)  # shuffles globally
+    .map(process_date)  # maps each element
+    .map(process_sample_to_np)  # maps each element
+)
+```
+
+```{code-cell} ipython3
+list(dataset)
+```
+
+Next we would like to tokenize the `review` field. LLM models operate on integers (encoded words) rather than raw strings. `AutoTokenizer` generic class ships `from_pretrained` method - accessor to models and tokenizers hosted on HF services.
+
+Let's use `bert-base-uncased`, a case-insensitive BERT-based transformers model.
+
+```{code-cell} ipython3
+tokenizer = AutoTokenizer.from_pretrained("bert-base-uncased")
+tokenizer
+```
+
+Transforming a single review string yields a dictionary with three keys. We're only interested in `input_ids` since that is the encoded review.
+
+```{code-cell} ipython3
+review = hf_train[0]["review"]
+pprint(review)
+print("\n", tokenizer(review).keys(), "\n")
+pprint(tokenizer(review)["input_ids"])
+```
+
+Plugging the selected transformer is as easy as before. We implement the `process_transformer` function and pass it to the `map` method.
+
+```{code-cell} ipython3
+def process_transformer(sample: dict) -> dict:
+    sample["review"] = np.array(tokenizer(sample["review"])["input_ids"])
+    return sample
+
+dataset = (
+    grain.MapDataset.source(hf_train)
+    .shuffle(seed=10)
+    .map(process_date)
+    .map(process_transformer)
+)
+```
+
+Now samples are less human- but more machine-friendly.
+
+```{code-cell} ipython3
+dataset[1]
+```
+
+Time to build our final pipeline! The pipeline doesn't need to be restricted to `shuffle` and `map`. Grain has a rich API and hands us multiple functionalities such as: `filter`, `random_map`, `repeat`.
```

Comment:
Suggest adding a link to the API docs for grain here.

Reply:
Done! Added a link to `grain` namespace API page.